In [6]:
# Import the modules
import numpy as np
import pandas as pd
from pathlib import Path
from sklearn.metrics import balanced_accuracy_score, confusion_matrix, classification_report

---

## Split the Data into Training and Testing Sets

### Step 1: Read the `lending_data.csv` data from the `Resources` folder into a Pandas DataFrame.

In [7]:
# Read the CSV file from the Resources folder into a Pandas DataFrame
df = pd.read_csv("../Resources/lending_data.csv")
df.head()

# Review the DataFrame
df.head()

FileNotFoundError: [Errno 2] No such file or directory: '../Resources/lending_data.csv'

### Step 2: Create the labels set (`y`)  from the “loan_status” column, and then create the features (`X`) DataFrame from the remaining columns.

In [ ]:
y = df["loan_status"]
X = df.drop(columns = "loan_status")
y.head()
X.head()

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  
0                 1       22800  
1                 0       13600  
2                 0       16100  
3                 1       22700  
4                 1       23000

In [ ]:
# Separate the data into labels and features
from sklearn.model_selection import train_test_split

# Separate the y variable, the labels

# Separate the X variable, the features
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=42)

In [ ]:
# Review the y variable Series
len(y_train)

58152

In [ ]:
# Review the X variable DataFrame
len(X_train)

58152

In [ ]:
# review Y
len(y_test)

NameError: name 'y_test' is not defined

In [4]:
# review Y
len(y_test)

NameError: name 'y_test' is not defined

In [ ]:
# review the df
len(df)

77536

### Step 3: Check the balance of the labels variable (`y`) by using the `value_counts` function.

In [ ]:
# Check the balance of our target values
y.value_counts()

loan_status
0    75036
1     2500
Name: count, dtype: int64

### Step 4: Split the data into training and testing datasets by using `train_test_split`.

In [ ]:
from sklearn.model_selection import train_test_split

# Separate the y variable, the labels

# Separate the X variable, the features
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=42)

---

## Create a Logistic Regression Model with the Original Data

###  Step 1: Fit a logistic regression model by using the training data (`X_train` and `y_train`).

In [ ]:
# Import the LogisticRegression module from SKLearn
from sklearn.linear_model import LogisticRegression

# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
classifier = LogisticRegression(random_state=1)
classifier

# Fit the model using training data
classifier.fit(X_train, y_train)

LogisticRegression(random_state=1)

### Step 2: Save the predictions on the testing data labels by using the testing feature data (`X_test`) and the fitted model.

In [ ]:
# Make a prediction using the testing data
predictions = classifier.predict(X_test)

### Step 3: Evaluate the model’s performance by doing the following:

* Calculate the accuracy score of the model.

* Generate a confusion matrix.

* Print the classification report.

In [ ]:
# Print the balanced_accuracy score of the model
from sklearn.metrics import accuracy_score

# Display the accuracy score for the test dataset.
accuracy_score(y_test, predictions)

0.9924680148576145

In [ ]:
# Generate a confusion matrix for the model
confusion_matrix(y_test, predictions)

array([[18699,    93],
       [   53,   539]], dtype=int64)

In [ ]:
# Print the classification report for the model
print(classification_report(y_test,predictions))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00     18792
           1       0.85      0.91      0.88       592

    accuracy                           0.99     19384
   macro avg       0.93      0.95      0.94     19384
weighted avg       0.99      0.99      0.99     19384



### Step 4: Answer the following question.

**Question:** How well does the logistic regression model predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** fair, because of a high accuracy score

---

## Predict a Logistic Regression Model with Resampled Training Data

### Step 1: Use the `RandomOverSampler` module from the imbalanced-learn library to resample the data. Be sure to confirm that the labels have an equal number of data points. 

In [ ]:
# Import the RandomOverSampler module form imbalanced-learn
from imblearn.over_sampling import RandomOverSampler

# Instantiate the random oversampler model
# # Assign a random_state parameter of 1 to the model
sampler = RandomOverSampler(random_state=1)

# Fit the original training data to the random_oversampler model
# X_train_sampled, y_train_sampled = oversampler.fit_resample(X_train_,y_train)
X_train_sampled, y_train_sampled = sampler.fit_resample(X_train, y_train)

In [ ]:
# Count the distinct values of the resampled labels data
resampled_value_counts = pd.Series(y_train_sampled).value_counts()
"Resampled Label Value Counts:\n", resampled_value_counts

('Resampled Label Value Counts:\n',
 loan_status
 0    56244
 1    56244
 Name: count, dtype: int64)

### Step 2: Use the `LogisticRegression` classifier and the resampled data to fit the model and make predictions.

In [ ]:
# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
from sklearn.linear_model import LogisticRegression
classifier_resampled = LogisticRegression(random_state=1)

# Fit the model using the resampled training data
classifier_resampled.fit(X_train_sampled, y_train_sampled)

# Make a prediction using the testing data
predictions_resampled = classifier_resampled.predict(X_test)

### Step 3: Evaluate the model’s performance by doing the following:

* Calculate the accuracy score of the model.

* Generate a confusion matrix.

* Print the classification report.

In [ ]:
# Print the balanced_accuracy score of the model 
accuracy_resampled = accuracy_score(y_test, predictions_resampled)
print("Accuracy Score:", accuracy_resampled)

Accuracy Score: 0.994428394552208


In [ ]:
# Generate a confusion matrix for the model
conf_matrix_resampled = confusion_matrix(y_test, predictions_resampled)
print("Confusion Matrix:")
print(conf_matrix_resampled)
# True Positives: 590
# True Negatives: 18686
# False Positives : 106
# False Negatives : 2

Confusion Matrix:
[[18686   106]
 [    2   590]]


In [ ]:
# Print the classification report for the model
print("Classification Report:")
print(classification_report(y_test, predictions_resampled))

Classification Report:
              precision    recall  f1-score   support

           0       1.00      0.99      1.00     18792
           1       0.85      1.00      0.92       592

    accuracy                           0.99     19384
   macro avg       0.92      1.00      0.96     19384
weighted avg       1.00      0.99      0.99     19384



### Step 4: Answer the following question

**Question:** How well does the logistic regression model, fit with oversampled data, predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** logistic regression has a score of 99.44%, indicating that the model is effective in making correct predictions.  I think this will be reliable in predicting healthy or high-risk loans.
weighted avg 1.00 0.99 0.99 19384